homeassistant/components/icloud.py

```diff
@@ -0,0 +1,946 @@
+                                    tempnow.minute * 60 -
+                                    tempnow.second)
+                if ((self._starttime.year > tempnow.year or
+                     self._starttime.month > tempnow.month or
+                     self._starttime.day > tempnow.day) and
+                        remainingseconds < 0):
+                    remainingseconds = 86400 + remainingseconds
+                self._remaining = (remainingdays * 1440 +
+                                   round(remainingseconds / 60, 0))
+        if endtime is None:
+            self._endtime = None
+            self._endtext = None
+        else:
+            self._endtime = datetime(endtime[1], endtime[2], endtime[3],
+                                     endtime[4], endtime[5], 0, 0,
+                                     self._tz)
+            self._endtext = self._endtime.strftime("%A %d %B %Y %H.%M.%S")
+            if current:
+                self._remaining = self._endtime - tempnow
+                remainingdays = self._remaining.days
+                remainingseconds = (self._endtime.hour * 3600 +
+                                    self._endtime.minute * 60 +
+                                    self._endtime.second -
+                                    tempnow.hour * 3600 -
+                                    tempnow.minute * 60 -
+                                    tempnow.second)
+                if ((self._endtime.year > tempnow.year or
+                     self._endtime.month > tempnow.month or
+                     self._endtime.day > tempnow.day) and
+                        remainingseconds < 0):
+                    remainingseconds = 86400 + remainingseconds
+                self._remaining = (remainingdays * 1440 +
+                                   round(remainingseconds / 60, 0))
+        self._duration = duration
+        self._title = title
+        if (current or nextev) and title is None:
+            self._title = 'Free'
+        self._location = location
+
+        tempdays = floor(self._remaining / 1440)
+        temphours = floor((self._remaining % 1440) / 60)
+        tempminutes = floor(self._remaining % 60)
+        self._remainingtext = (str(tempdays) + "d " +
+                               str(temphours) + "h " +
+                               str(tempminutes) + "m")
+
+        if self._remaining <= 0:
+            self.hass.states.remove(self.entity_id)
+        else:
+            self.update_ha_state()
+
+    def check_alive(self):
+        """Check if event is over."""
+        current = self._type == TYPE_CURRENT
+        nextev = self._type == TYPE_NEXT
+        self._remaining = 0
+        tempnow = dt_util.now(self._tz)
+        if self._starttime is not None:
+            if nextev:
+                self._remaining = self._starttime - tempnow
+                remainingdays = self._remaining.days
+                remainingseconds = (self._starttime.hour * 3600 +
```

Comment:
That's the thing, when I compared the datetimes, the remaining time was always 45 minutes off. :s